File: ex20/ex20.cpp
#include <bits/stdc++.h>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        int n = s.size();
        if (n % 2 == 1)
            return false;
        stack<char> st;
        unordered_map<char, char> hashtable = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
        for (auto &c : s)
        {
            if (hashtable.count(c))
            {
                st.push(hashtable[c]);
            }
            else
            {
                if (st.empty() || st.top() != c)
                    return false;
                st.pop();
            }
        }
        return st.empty();

    }
};

int main(void)
{

    return 0;
}

File: ex20/ex20.py
#!/usr/bin/env python3

class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        stack = list()
        pairs = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        for ch in s:
            if ch in pairs:
                stack.append(pairs[ch])
            else:
                if not stack or stack[-1] != ch:
                    return False
                stack.pop()
        return not stack


class Solution2:
    def isValid(self, s: str) -> bool:
        st = []
        for i in s:
            if i == '(' or i == '{' or i == '[':
                st.append(i)
            else:
                if i == ')' and len(st) and st[-1] == '(':
                    st.pop()
                elif i == '}' and len(st) and st[-1] == '{':
                    st.pop()
                elif i == ']' and len(st) and st[-1] == '[':
                    st.pop()
                else:
                    return False
        return False if len(st) else True


sy = Solution()
print(sy.isValid('(()'))
print(sy.isValid('(()[]){}'))
print(sy.isValid('(['))
